util/expensivequery/memory_usage_alarm.go

```diff
@@ -44,9 +59,66 @@ type memoryUsageAlarm struct {
 	tmpDir              string
 	lastLogFileName     []string
 	lastProfileFileName [][]string // heap, goroutine
+
+	s3Conf s3Config
+}
+
+func (record *memoryUsageAlarm) initS3Config() {
+	if recordToS3 := config.GetGlobalConfig().S3.EnableUploadOOMRecord; recordToS3 {
+		record.s3Conf.accessKey = config.GetGlobalConfig().S3.AccessKey
+		record.s3Conf.secretKey = config.GetGlobalConfig().S3.SecretKey
+		record.s3Conf.endPoint = config.GetGlobalConfig().S3.EndPoint
+		record.s3Conf.bucketName = config.GetGlobalConfig().S3.BucketName
+		record.s3Conf.disableSSL = config.GetGlobalConfig().S3.DisableSSL
+		record.s3Conf.s3ForcePathStyle = config.GetGlobalConfig().S3.S3ForcePathStyle
+		record.s3Conf.enableUploadOOMRecord = config.GetGlobalConfig().S3.EnableUploadOOMRecord
+	}
+}
+
+func (record *memoryUsageAlarm) uploadFileToS3(filenames []string) {
+	if record.s3Conf.enableUploadOOMRecord && record.initialized && time.Since(record.lastCheckTime) > 600*time.Second {
+		sess, err := session.NewSession(&aws.Config{
+			Credentials:      credentials.NewStaticCredentials(record.s3Conf.accessKey, record.s3Conf.secretKey, ""),
+			Endpoint:         aws.String(record.s3Conf.endPoint),
+			Region:           aws.String(record.s3Conf.regionName),
+			DisableSSL:       aws.Bool(record.s3Conf.disableSSL),
+			S3ForcePathStyle: aws.Bool(record.s3Conf.s3ForcePathStyle),
+		})
+		if err != nil {
+			logutil.BgLogger().Error("create s3 new session fail", zap.Error(err))
+			return
+		}
+		uploader := s3manager.NewUploader(sess)
+
+		for _, filename := range filenames {
+			file, err := os.OpenFile(filename, os.O_RDONLY, 0600)
```

Comment:
defer in the loop is bad. I think you can refactor the code block in the loop into a function.

Reply:
> defer in the loop is bad. I think you can refactor the code block in the loop into a function.

done.